File: 23/2.cpp
#include <bits/stdc++.h>
#define N 1000000
#define M 10000000

using namespace std;

struct Node {
	Node *next;
	int el;
	Node(int el, Node *next) {
		if(next == nullptr)
			this->next = this;
		else {
			this->next = next;
		}
		this->el = el;
	}
	Node* append(int el) {
		next = new Node(el, next);
		return next;
	}
};

Node **nodes = new Node*[N+1];

int main() {
	vector<int> input;
	string s;
	cin >> s;
	for(auto u:s) {
		input.push_back(u-'0');
	}
	for(int i=0; i<input.size(); i++) {
		if(i == 0) {
			nodes[input[i]] = new Node(input[i], nullptr);
		} else {
			nodes[input[i]] = nodes[input[i-1]]->append(input[i]);
		}
	}
	nodes[input.size()+1] = nodes[input[input.size()-1]]->append(input.size()+1);
	for(int i=input.size()+2; i<=N; i++)
		nodes[i] = nodes[i-1]->append(i);
	Node *cur = nodes[input[0]];
	for(int i=0; i<M; i++) {
		Node *s = cur->next;
		Node *e = cur->next->next->next;
		cur->next = e->next;
		int dest = cur->el;
		do {
			dest--;
			if(dest == 0)
				dest = N;
		} while(dest == s->el || dest == s->next->el || dest == e->el);
		e->next = nodes[dest]->next;
		nodes[dest]->next = s;
		cur = cur->next;
	}
	cur = nodes[1];
	cout << (((long long)cur->next->el)*(cur->next->next->el));
}

File: 23/2.py
import sys

class Node:
    def __init__(self, el, next):
        if next == None:
            self.next = self
        else:
            self.next = next
        self.el = el
    def append(self, el):
        self.next = Node(el, self.next)
        return self.next

input = list(map(int, sys.stdin.readline().strip()))
for i in range(len(input)+1, 1000000+1):
    input.append(i)

c_n = [None]*(len(input)+1)

for i in range(len(input)):
    if i == 0:
        c_n[input[i]] = Node(input[i], None)
    else:
        c_n[input[i]] = c_n[input[i-1]].append(input[i])

cur = c_n[input[0]]

for i in range(10000000):
    s = cur.next
    e = cur.next.next.next
    cur.next = e.next
    dest = cur.el-1
    if dest == 0:
        dest = len(input)
    while dest == s.el or dest == s.next.el or dest == s.next.next.el:
        dest = dest-1
        if dest == 0:
            dest = len(input)
    e.next = c_n[dest].next
    c_n[dest].next = s
    cur = cur.next

cur = c_n[1]
print(cur.next.el*cur.next.next.el)
